File: p41.py

"""Final project, part 1"""
import numpy as np
import matplotlib.pyplot as plt
import m2
fl = m2.flow
import time
#from m2 import flow as fl #assumes p41.f90 has been compiled with: f2py -c p41.f90 -m m2


def jacobi(n,kmax=10000,tol=1.0e-8,s0=0.1,display=False):
    """ Solve liquid flow model equations with
        jacobi iteration.
        Input:
            n: number of grid points in r and theta
            kmax: max number of iterations
            tol: convergence test parameter
            s0: amplitude of cylinder deformation
            display: if True, plots showing the velocity field and boundary deformation
            are generated
        Output:
            w,deltaw: Final velocity field and |max change in w| each iteration
    """

    #-------------------------------------------
    #Set Numerical parameters and generate grid
    Del_t = 0.5*np.pi/(n+1)
    Del_r = 1.0/(n+1)
    Del_r2 = Del_r**2
    Del_t2 = Del_t**2
    r = np.linspace(0,1,n+2)
    t = np.linspace(0,np.pi/2,n+2) #theta
    tg,rg = np.meshgrid(t,r) # r-theta grid

    #Factors used in update equation (after dividing by gamma)
    rg2 = rg*rg
    fac = 0.5/(rg2*Del_t2 + Del_r2)
    facp = rg2*Del_t2*fac*(1+0.5*Del_r/rg) #alpha_p/gamma
    facm = rg2*Del_t2*fac*(1-0.5*Del_r/rg) #alpha_m/gamma
    fac2 = Del_r2*fac #beta/gamma
    RHS = fac*(rg2*Del_r2*Del_t2) #1/gamma

    #set initial condition/boundary deformation
    w0 = (1-rg**2)/4 #Exact solution when s0=0
    s_bc = s0*np.exp(-10.*((t-np.pi/2)**2))/Del_r
    fac_bc = s_bc/(1+s_bc)

    deltaw = []
    w = w0.copy()
    wnew = w0.copy()

    #Jacobi iteration
    for k in range(kmax):
        #Compute wnew
        wnew[1:-1,1:-1] = RHS[1:-1,1:-1] + w[2:,1:-1]*facp[1:-1,1:-1] + w[:-2,1:-1]*facm[1:-1,1:-1] + (w[1:-1,:-2] + w[1:-1,2:])*fac2[1:-1,1:-1] #Jacobi update

        #Apply boundary conditions
        wnew[:,0] = wnew[:,1] #theta=0
        wnew[:,-1] = wnew[:,-2] #theta=pi/2
        wnew[0,:] = wnew[1,:] #r=0
        wnew[-1,:] = wnew[-2,:]*fac_bc #r=1s

        #Compute delta_p
        deltaw += [np.max(np.abs(w-wnew))]
        w = wnew.copy()
        #if k%1000==0: print("k,dwmax:",k,deltaw[k])
        #check for convergence
        if deltaw[k]<tol:
            print("Converged,k=%d,dw_max=%28.16f " %(k,deltaw[k]))
            break

    deltaw = deltaw[:k+1]

    if display:
        #plot final velocity field, difference from initial guess, and cylinder
        #surface
        plt.figure()
        plt.contour(t,r,w,50)
        plt.xlabel(r'$\theta$')
        plt.ylabel('r')
        plt.title('Final velocity field, python jacobi')
        plt.savefig('velfield_pj.png')
        plt.show()

        plt.figure()
        plt.contour(t,r,np.abs(w-w0),50)
        plt.xlabel(r'$\theta$')
        plt.ylabel('r')
        plt.title(r'$|w - w_0|$, python jacobi')
        plt.savefig('ww0_diff_pj.png')
        plt.show()

        plt.figure()
        plt.polar(t,np.ones_like(t),'k--')
        plt.polar(t,np.ones_like(t)+s_bc*Del_r,'r-')
        plt.title('Deformed cylinder surface, python jacobi')
        plt.savefig('cylsurface_pj.png')
        plt.show()

    return w,deltaw



def performance(n,nrange,nrange2):

    """ PLOT PYTHON JACOBI DATA  """
    jacobi(n,display=True)

    #set up variables for plotting fortran jacobi data
    """ PLOT FORTRAN JACOBI DATA & SGISOLVE VECTORFIELD """
    del_t = 0.5*np.pi/(n+1)
    del_r = 1.0/(n+1)
    del_r2 = del_r**2
    del_t2 = del_t**2
    r = np.linspace(0,1,n+2)
    t = np.linspace(0,np.pi/2,n+2) #theta
    tg,rg = np.meshgrid(t,r) # r-theta grid

    w0_j = (1-rg**2)/4 #Exact solution when s0=0
    s_bc = fl.fl_s0*np.exp(-10.*((t-np.pi/2)**2))/del_r
    fac_bc = s_bc/(1+s_bc)

    w_j = fl.jacobi(n) #vector field returned by Fortran Jacobi

    s,fac,facm,facp,fac2,fac_bc = np.zeros(n+2),np.zeros(n+2),np.zeros(n+2),np.zeros(n+2),np.zeros(n+2),np.zeros(n+2)

    s = np.zeros(n+2)
    for s1 in range(n+2):
        s[s1] = fl.fl_s0*(np.exp(-10*(s1*del_t - (np.pi/2.0))**2) + np.exp(-10*(s1*del_t + (np.pi/2.0))**2))
    for p2 in range(1,n+1):
        fac[p2] = 1.0/((2.0/del_r2) + (2.0/(del_t2*(p2*del_r)**2)))
    for p1 in range(1,n+1):
        facm[p1] = (1.0/del_r2 - 1.0/(2.0*p1*del_r2))*fac[p1]
        facp[p1] = (1.0/del_r2 + 1.0/(2.0*p1*del_r2))*fac[p1]
        fac2[p1] = (1.0/(del_t2*(p1*del_r)**2))*fac[p1]
    fac_bc = s/(del_r + s)

    w_unpk = fl.sgisolve(n) # unpacked vector field returned by Fortran sgisolve
    w_pk = np.zeros((n+2,n+2))

    for i in range(n+2):
        w_pk[i,:] = w_unpk[i*(n+2):(i+1)*(n+2)]  #packing the Fortran sgisolve vector field.

    plt.figure()
    plt.contour(t,r,w_j,50)
    plt.xlabel(r'$\theta$')
    plt.ylabel('r')
    plt.title('Final velocity field, fortran jacobi')
    plt.savefig('velfield_fj.png')
    plt.show()

    plt.figure()
    plt.contour(t,r,w_pk,50)
    plt.xlabel(r'$\theta$')
    plt.ylabel('r')
    plt.title('Final velocity field, fortran sgisolve')
    plt.savefig('velfield_fsgi.png')
    plt.show()

    plt.figure()
    plt.contour(t,r,np.abs(w_j-w0_j),50)
    plt.xlabel(r'$\theta$')
    plt.ylabel('r')
    plt.title(r'$|w - w_0|$, fortran jacobi')
    plt.savefig('ww0_diff_fj.png')
    plt.show()



    plt.figure()
    plt.polar(t,np.ones_like(t),'k--')
    plt.polar(t,np.ones_like(t)+s_bc*del_r,'r-')
    plt.title('Deformed cylinder surface, fortran jacobi')
    plt.savefig('cylsurface_fj.png')
    plt.show()

    """ NOW WE LOOK AT THE SPEEDS OF THE FORTRAN SUBROUTINES MVEC & MTVEC AGAINST N  """

    labels1 = ['mvec','mtvec','jacobi_fortran','jacobi_python','sgisolve(numthreads=2)','sgisolve(numthreads=1)']
    times = np.zeros((len(labels1),len(nrange)))
    j = 0
    for n_i in nrange:
        fl.numthreads = 2
        s,fac,facm,facp,fac2,fac_bc = np.zeros(n_i+2),np.zeros(n_i+2),np.zeros(n_i+2),np.zeros(n_i+2),np.zeros(n_i+2),np.zeros(n_i+2)
        #set up coefficient vectors
        del_t = 0.5*np.pi/(n_i+1)
        del_r = 1.0/(n_i+1)
        del_r2 = del_r**2
        del_t2 = del_t**2
        r = np.zeros((n_i+2,n_i+2))
        t = np.linspace(0,np.pi/2,n+2) #theta

        s = np.zeros(n_i+2)
        for s1 in range(n_i+2):
            s[s1] = fl.fl_s0*(np.exp(-10*(s1*del_t - (np.pi/2.0))**2) + np.exp(-10*(s1*del_t + (np.pi/2.0))**2))
        for p2 in range(1,n_i+1):
            fac[p2] = 1.0/((2.0/del_r2) + (2.0/(del_t2*(p2*del_r)**2)))
        for p1 in range(1,n_i+1):
            facm[p1] = (1.0/del_r2 - 1.0/(2.0*p1*del_r2))*fac[p1]
            facp[p1] = (1.0/del_r2 + 1.0/(2.0*p1*del_r2))*fac[p1]
            fac2[p1] = (1.0/(del_t2*(p1*del_r)**2))*fac[p1]
        fac_bc = s/(del_r + s)

        #first generated a vector w to input into mvec & mtvec
        w_packed = fl.jacobi(n_i)
        w = np.zeros((n_i+2)**2)
        for i in range(n_i+2):
            w[i*(n_i+2):(i+1)*(n_i+2)] = w_packed[i,:] #unpacks w_packed into a single 1D array
            r[i,:] = i*del_r

        mv1 = time.time()  #mvec time
        mv_out = fl.mvec(fac,fac2,facp,facm,fac_bc,w,n_i)
        mv2 = time.time()
        times[0,j] = mv2 - mv1

        mt1 = time.time()   #mtvec time
        mt_out = fl.mtvec(fac,fac2,facp,facm,fac_bc,w,n_i)
        mt2 = time.time()
        times[1,j] = mt2 - mt1

        jf1 = time.time()      #jacobi fortran time
        w_jf = fl.jacobi(n_i)
        jf2 = time.time()
        times[2,j] = jf2-jf1

        jp1 = time.time()       #jacobi python time
        w_jp = jacobi(n_i,display=False)
        jp2 = time.time()
        times[3,j] = jp2-jp1

        sgi1 = time.time()      #sgisolve time when numthreads = 2
        w_s = fl.sgisolve(n_i)
        sgi2 = time.time()
        times[4,j] = sgi2-sgi1

        fl.numthreads = 1

        sgi3 = time.time()      #sgisolve time when numthreads = 1
        w_s = fl.sgisolve(n_i)
        sgi4 = time.time()
        times[5,j] = sgi4-sgi3

        j += 1

    labels2 = ['mvec','mtvec','jacobi_fortran','jacobi_python','sgisolve(numthreads=2)','sgisolve(numthreads=1)']
    times2 = np.zeros((len(labels2),len(nrange2)))
    j = 0
    for n_i in nrange2:
        fl.numthreads = 2
        s,fac,facm,facp,fac2,fac_bc = np.zeros(n_i+2),np.zeros(n_i+2),np.zeros(n_i+2),np.zeros(n_i+2),np.zeros(n_i+2),np.zeros(n_i+2)
        #set up coefficient vectors
        del_t = 0.5*np.pi/(n_i+1)
        del_r = 1.0/(n_i+1)
        del_r2 = del_r**2
        del_t2 = del_t**2
        r = np.zeros((n_i+2,n_i+2))
        t = np.linspace(0,np.pi/2,n+2) #theta

        s = np.zeros(n_i+2)
        for s1 in range(n_i+2):
            s[s1] = fl.fl_s0*(np.exp(-10*(s1*del_t - (np.pi/2.0))**2) + np.exp(-10*(s1*del_t + (np.pi/2.0))**2))
        for p2 in range(1,n_i+1):
            fac[p2] = 1.0/((2.0/del_r2) + (2.0/(del_t2*(p2*del_r)**2)))
        for p1 in range(1,n_i+1):
            facm[p1] = (1.0/del_r2 - 1.0/(2.0*p1*del_r2))*fac[p1]
            facp[p1] = (1.0/del_r2 + 1.0/(2.0*p1*del_r2))*fac[p1]
            fac2[p1] = (1.0/(del_t2*(p1*del_r)**2))*fac[p1]
        fac_bc = s/(del_r + s)

        #first generated a vector w to input into mvec & mtvec
        w_packed = fl.jacobi(n_i)
        w = np.zeros((n_i+2)**2)
        for i in range(n_i+2):
            w[i*(n_i+2):(i+1)*(n_i+2)] = w_packed[i,:] #unpacks w_packed into a single 1D array
            r[i,:] = i*del_r

        mv1 = time.time()  #mvec time
        mv_out = fl.mvec(fac,fac2,facp,facm,fac_bc,w,n_i)
        mv2 = time.time()
        times2[0,j] = mv2 - mv1

        mt1 = time.time()   #mtvec time
        mt_out = fl.mtvec(fac,fac2,facp,facm,fac_bc,w,n_i)
        mt2 = time.time()
        times2[1,j] = mt2 - mt1

        jf1 = time.time()      #jacobi fortran time
        w_jf = fl.jacobi(n_i)
        jf2 = time.time()
        times2[2,j] = jf2-jf1

        jp1 = time.time()       #jacobi python time
        w_jp = jacobi(n_i,display=False)
        jp2 = time.time()
        times2[3,j] = jp2-jp1

        sgi1 = time.time()      #sgisolve time when numthreads = 2
        w_s = fl.sgisolve(n_i)
        sgi2 = time.time()
        times2[4,j] = sgi2-sgi1

        fl.numthreads = 1

        sgi3 = time.time()      #sgisolve time when numthreads = 1
        w_s = fl.sgisolve(n_i)
        sgi4 = time.time()
        times2[5,j] = sgi4-sgi3

        j += 1

    plt.figure()
    for i in range(1):
        plt.plot(nrange,times[i,:],label=labels1[i])
    plt.legend()
    plt.xlabel('n')
    plt.ylabel('runtime, s')
    plt.title('Runtime of the Fortran subroutine mvec against n')
    plt.savefig('mvec_vs_n.png')
    plt.show()

    plt.figure()
    for i in range(2):
        plt.plot(nrange,times[i,:],label=labels1[i])
    plt.legend()
    plt.xlabel('n')
    plt.ylabel('runtimes, s')
    plt.title('Runtimes of the Fortran subroutines mvec & mtvec against n')
    plt.savefig('mvecmtvec_vs_n.png')
    plt.show()

    plt.figure()
    for i in range(2,len(labels1)):
        plt.plot(nrange,times[i,:],label=labels1[i])
    plt.legend()
    plt.xlabel('n')
    plt.ylabel('runtimes, s')
    plt.title('Runtimes of the Fortran subroutines & python jacobi against n')
    plt.savefig('sgi&jacvs_n.png')
    plt.show()

    plt.figure()
    for i in range(2,len(labels2)):
        plt.plot(nrange2,times2[i,:],label=labels2[i])
    plt.legend()
    plt.xlabel('n')
    plt.ylabel('runtimes, s')
    plt.title('Runtimes of the Fortran subroutines & python jacobi against n')
    plt.savefig('sgi&jacvs_n_long.png')
    plt.show()

    return None



if __name__=='__main__':

    n = 30
    print("n = ",n)
    fl.fl_s0, fl.numthreads= 0.1, 0.2
    nrange, nrange2 = np.arange(1,37,5), np.arange(1,72,10)
    print('nrange = ',nrange)
    print('nrange2 = ',nrange2)
    performance(n,nrange,nrange2)

File: p41.f90
!M3C FINAL PROJECT
!Final project part 1
!Module for flow simulations of liquid through tube
!This module contains a few module variables (see comments below)
!and four subroutines:
!jacobi: Uses jacobi iteration to compute solution
! to flow through tube
!sgisolve: To be completed. Use sgi method to
! compute flow through tube
!mvec: To be completed; matrix-vector multiplication z = Ay
!mtvec: To be completed; matrix-vector multiplication z = A^T y
!--------------------------------------------------
!p41.f90 COMPILED AS m2 USING THE UNIX COMMAND:
!   python -m numpy.f2py --f90flags='-fopenmp' -c p41_template.f90 -m m2 -lgomp
!--------------------------------------------------
module flow
    use omp_lib
    implicit none
    real(kind=8), parameter :: pi = acos(-1.d0)
    integer :: numthreads !number of threads used in parallel regions
    integer :: fl_kmax=10000 !max number of iterations
    real(kind=8) :: fl_tol=0.00000001d0 !convergence criterion
    real(kind=8), allocatable, dimension(:) :: fl_deltaw !|max change in w| each iteration
    real(kind=8) :: fl_s0=0.1d0 !deformation magnitude

contains
!-----------------------------------------------------
!Solve 2-d tube flow problem with Jacobi iteration
subroutine jacobi(n,w)
    !input  n: number of grid points (n+2 x n+2) grid
    !output w: final velocity field
    !Should also compute fl_deltaw(k): max(|w^k - w^k-1|)
    !A number of module variables can be set in advance.

    integer, intent(in) :: n
    real(kind=8), dimension(0:n+1,0:n+1), intent(out) :: w
    integer :: i1,j1,k1
    real(kind=8) :: del_r,del_t,del_r2,del_t2
    real(kind=8), dimension(0:n+1) :: s_bc,fac_bc
    real(kind=8), dimension(0:n+1,0:n+1) :: r,r2,t,RHS,w0,wnew,fac,fac2,facp,facm

    if (allocated(fl_deltaw)) then
      deallocate(fl_deltaw)
    end if
    allocate(fl_deltaw(fl_kmax))


    !grid--------------
    del_t = 0.5d0*pi/dble(n+1)
    del_r = 1.d0/dble(n+1)
    del_r2 = del_r**2
    del_t2 = del_t**2


    do i1=0,n+1
        r(i1,:) = i1*del_r
    end do

    do j1=0,n+1
        t(:,j1) = j1*del_t
    end do
    !-------------------

    !Update-equation factors------
    r2 = r**2
    fac = 0.5d0/(r2*del_t2 + del_r2)
    facp = r2*del_t2*fac*(1.d0+0.5d0*del_r/r) !alpha_p/gamma
    facm = r2*del_t2*fac*(1.d0-0.5d0*del_r/r) !alpha_m/gamma
    fac2 = del_r2 * fac !beta/gamma
    RHS = fac*(r2*del_r2*del_t2) !1/gamma
    !----------------------------

    !set initial condition/boundary deformation
    w0 = (1.d0-r2)/4.d0
    w = w0
    wnew = w0
    s_bc = fl_s0*exp(-10.d0*((t(0,:)-pi/2.d0)**2))/del_r
    fac_bc = s_bc/(1.d0+s_bc)


    !Jacobi iteration
    do k1=1,fl_kmax
        wnew(1:n,1:n) = RHS(1:n,1:n) + w(2:n+1,1:n)*facp(1:n,1:n) + w(0:n-1,1:n)*facm(1:n,1:n) + &
                                         (w(1:n,0:n-1) + w(1:n,2:n+1))*fac2(1:n,1:n)

        !Apply boundary conditions
        wnew(:,0) = wnew(:,1) !theta=0
        wnew(:,n+1) = wnew(:,n) !theta=pi/2
        wnew(0,:) = wnew(1,:) !r=0
        wnew(n+1,:) = wnew(n,:)*fac_bc !r=1s

        fl_deltaw(k1) = maxval(abs(wnew-w)) !compute relative error

        w=wnew    !update variable
        if (fl_deltaw(k1)<fl_tol) exit !check convergence criterion
        if (mod(k1,1000)==0) print *, k1,fl_deltaw(k1)
    end do

    print *, 'k,error=',k1,fl_deltaw(min(k1,fl_kmax))

end subroutine jacobi
!-----------------------------------------------------

!Solve 2-d tube flow problem with sgi method
subroutine sgisolve(n,w)
  implicit none
  !input  n: number of grid points (n+2 x n+2) grid
  !output w: final velocity field stored in a column vector
  !Should also compute fl_deltaw(k): max(|w^k - w^k-1|)
  !A number of module variables can be set in advance.
  integer, intent(in) :: n
  real(kind=8), dimension((n+2)*(n+2)), intent(out) :: w
  real(kind=8) :: del_t,del_r,k,mu,del_r2,del_t2
  real(kind=8), allocatable, dimension(:,:) :: M
  real(kind=8), dimension((n+2)*(n+2)) :: b,v,d,x,e,enew,A_d,dnew,M_d,xnew
  real(kind=8), dimension(0:n+1) :: fac2,facm,facp,fac_bc,fac
  real(kind=8), dimension(0:n+1) :: s,s_j
  integer :: l1,p1,s1,b1,p2
  !add other variables as needed
  if (allocated(fl_deltaw)) then
    deallocate(fl_deltaw)
  end if
  allocate(fl_deltaw(fl_kmax))

  !-------------------------------------------
  !WE USE OPENMP TO PARALLISE THE SET UP OF THE COEFFICIENT VECTORS.
  !AS MVEC & MTVEC ARE REPEATEDLY USED IN THE MAIN LOOP, WE CANNOT PARALLELISE
  !THESE SUBROUTINES AS THIS WOULD CREATE PARALLEL REGIONS WITHIN PARALLEL
  !REGIONS AND CRITICAL VARAIABLE INFORMATION WOULD BE LOST.
  !-------------------------------------------

  !$ call omp_set_num_threads(numthreads)

  del_t = 0.5d0*pi/dble(n+1)  !grid spacings
  del_r = 1.d0/dble(n+1)
  del_t2 = del_t**2
  del_r2 = del_r**2

  !now set up coefficient vectors

  !$OMP parallel do private(s1)
  do s1 = 0,n+1
    s(s1) = fl_s0*(exp(-10*(s1*del_t - (pi/2.d0))**2) + exp(-10*(s1*del_t + (pi/2.d0))**2))
  end do
  !$OMP end parallel do

  !$OMP parallel do private(p2)
  do p2 = 1,n
    fac(p2) = 1.d0/((2.d0/del_r2) + (2.d0/(del_t2*(p2*del_r)**2)))
  end do
  !$OMP end parallel do

  !$OMP parallel do private(p1)
  do p1 = 1,n
    facm(p1) = (1.d0/del_r2 - 1.d0/(2.d0*p1*del_r2))*fac(p1)
    facp(p1) = (1.d0/del_r2 + 1.d0/(2.d0*p1*del_r2))*fac(p1)
    fac2(p1) = (1.d0/(del_t2*(p1*del_r)**2))*fac(p1)
  end do
  !$OMP end parallel do

  fac_bc = s/dble(del_r + s)

  b = 0.d0
  !$OMP parallel do private(b1)
  do b1 = 1,n
    b((b1*(n+2))+2:(b1*(n+2))+n+1) = -1.d0*fac(b1)
  end do
!$OMP end parallel do

  !initial states
  call mtvec(n,fac,fac2,facp,facm,fac_bc,b,v) !compute v

  d = v
  x = 0.d0
  e = v
  !main loop
  !------------------------------------------------------
  !------------------------------------------------------
  !THE MAIN ITERATIVE LOOP OF SGISOLVE IS NOT THREADSAFE, SO
  !IT IS NOT POSSIBLE TO PARALLISE EFFICIENTLY.
  !------------------------------------------------------
  !------------------------------------------------------
  do l1 = 1,fl_kmax
    call mvec(n,fac,fac2,facp,facm,fac_bc,d,A_d) !computes matrix vector product A d
    call mtvec(n,fac,fac2,facp,facm,fac_bc,A_d,M_d) !computes mvecprod Md = A^T A d
    k = dble(dot_product(e,e))/dble(dot_product(d,M_d))

    xnew = x + k*d
    enew = e - k*M_d
    mu = dble(dot_product(enew,enew))/dble(dot_product(e,e))
    dnew = enew + mu*d

    fl_deltaw(l1) = maxval(abs(enew-e))

    if (fl_deltaw(l1)<fl_tol) exit !check convergence criterion
    if (mod(l1,1000)==0) print *, l1,fl_deltaw(l1)

    !update variables
    e = enew
    d = dnew
    x = xnew

  end do
  w = x

end subroutine sgisolve

!------------------------------------------------------
!------------------------------------------------------

!Compute matrix-vector multiplication, z = Ay
subroutine mvec(n,fac,fac2,facp,facm,fac_bc,y,z)
    !input n: grid is (n+2) x (n+2)
    ! fac,fac2,facp,facm,fac_bc: arrays that appear in
    !   discretized equations
    ! y: vector multipled by A
    !output z: result of multiplication Ay
    implicit none
    integer, intent(in) :: n
    real(kind=8), dimension(n+2), intent(in) :: fac,fac2,facp,facm,fac_bc
    real(kind=8), dimension((n+2)*(n+2)), intent(in) :: y
    real(kind=8), dimension((n+2)*(n+2)), intent(out) :: z
    integer :: i1,j1,j2,i2,k

    !From the governing equation
    do i1 = 1,n
      do j1 = 2,n+1
        k = i1*(n+2) + j1
        z(k) = y(k-(n+2))*facm(i1+1) + y(k-1)*fac2(i1+1) + y(k)*(-1.d0) + y(k+1)*fac2(i1+1) + y(k+n+2)*facp(i1+1)
      end do
    end do

    !Boundary conditions
    do i2 = 1,n
      z((i2*(n+2))+1) = y((i2*(n+2))+1)*(-1.d0) + 1.d0*y((i2*(n+2))+2)  !for w_i,1 - w_i,0 = 0 in row i,0
      z((i2*(n+2))+n+2) = y((i2*(n+2))+n+2)*(1.d0) + (-1.d0)*y((i2*(n+2))+n+1)  !for w_i,n+1 - w_i,n = 0 in row i,n+1
    end do
    do j2 = 1,(n+2)
      z(j2) = y(n+2+j2) + (-1.d0)*y(j2) !w_1,j - w0,j = 0 in row 0,j
      z(((n+1)*(n+2))+j2) = 1.d0*y(((n+1)*(n+2))+j2) + (-1.d0)*fac_bc(j2)*y((n*(n+2))+j2)  ! for w_n+1,j - w_n,j*fac_bc(j) in row w_n+1,j
    end do

end subroutine mvec

!------------------------------------------------------
!------------------------------------------------------


!------------------------------------------------------
!------------------------------------------------------


!Compute matrix-vector multiplication, z = A^T y
subroutine mtvec(n,fac,fac2,facp,facm,fac_bc,y,z)
    !input n: grid is (n+2) x (n+2)
    ! fac,fac2,facp,facm,fac_bc: arrays that appear in
    !   discretized equations
    ! y: vector multipled by A^T
    !output z: result of multiplication A^T y
    implicit none
    integer, intent(in) :: n
    real(kind=8), dimension(n+2), intent(in) :: fac,fac2,facp,facm,fac_bc
    real(kind=8), dimension((n+2)*(n+2)), intent(in) :: y
    real(kind=8), dimension((n+2)*(n+2)), intent(out) :: z
    real(kind=8), dimension((n+2)*(n+2)) :: s,s1,finalvec
    !real(kind=8), dimension((n+2)**2,(n+2)**2) :: A,A_t
    integer :: i1,j1,j2,i2,k,k2,k3,k4

    !add other variables as needed
    finalvec = 0.d0

    !Governing equations

    do i1 = 1,n
      do j1 = 2,n+1
        k = i1*(n+2) + j1
        s = 0.d0

        s(k-(n+2)) = facm(i1+1)
        s(k-1) = fac2(i1+1)
        s(k) = -1.d0
        s(k+1) = fac2(i1+1)
        s(k+n+2) = facp(i1+1)

        finalvec = finalvec + y(k)*s
      end do
    end do

    !Boundary conditions
    do i2 = 1,n
      s = 0.d0
      s1 = 0.d0
      k2 = (i2*(n+2))+1

      s(k2) = -1.d0
      s(k2+1) = 1.d0

      finalvec = finalvec + y(k2)*s

      k3 = (i2*(n+2))+n+2

      s1(k3) = 1.d0
      s1(k3-1) = -1.d0

      finalvec = finalvec + y(k3)*s1
    end do

    do j2 = 1,(n+2)
      s = 0.d0
      s1 = 0.d0

      s(n+2+j2) = 1.d0  !w_1,j - w0,j = 0 in row 0,j
      s(j2) = -1.d0
      finalvec = finalvec + y(j2)*s

      k4 = ((n+1)*(n+2))+j2
      s1(k4) = 1.d0 ! for w_n+1,j - w_n,j*fac_bc(j) in row w_n+1,j
      s1(k4-(n+2)) = -1.d0*fac_bc(j2)
      finalvec = finalvec + y(k4)*s1
    end do

    z = finalvec

end subroutine mtvec

end module flow
